syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.github.yzernik.squeakserver";
option java_outer_classname = "SqueakProto";
option objc_class_prefix = "SQK";

package squeakserver;

// Interface exported by the server.
service SqueakServer {

  /** sqk: `postsqueak`
  */
  rpc PostSqueak (PostSqueakRequest) returns (PostSqueakReply) {}

  /** sqk: `getsqueak`
  */
  rpc GetSqueak (GetSqueakRequest) returns (GetSqueakReply) {}

  /** sqk: `lookupsqueaks`
  */
  rpc LookupSqueaks (LookupSqueaksRequest) returns (LookupSqueaksReply) {}

  /** sqk: `buysqueak`
  */
  rpc BuySqueak (BuySqueakRequest) returns (BuySqueakReply) {}

}

message PostSqueakRequest {
    /// Content of the squeak to be made.
    Squeak squeak = 1;
}

message GetSqueakRequest {
    /// Hash of the squeak to get.
    bytes hash = 1;
}

message LookupSqueaksRequest {
    /// List of the author addresses
    repeated string addresses = 1;
    /// Minimum block to look for
    int32 min_block = 2;
    /// Maximum block to look for
    int32 max_block = 3;
}

message BuySqueakRequest {
    /// Hash of the squeak to buy.
    bytes hash = 1;
}

message PostSqueakReply {
}

message GetSqueakReply {
    /// The squeak.
    Squeak squeak = 1;
}

message LookupSqueaksReply {
    /// Hash of the squeak to get.
    repeated bytes hashes = 1;
}

message BuySqueakReply {
    /// The buy offer
    SqueakBuyOffer offer = 1;
}

message Squeak {
    /// Hash of the squeak.
    bytes hash = 1;

    /// Serialized squeak.
    bytes serialized_squeak = 2;
}

message SqueakBuyOffer {
    /// The squeak hash.
    bytes squeak_hash = 1;

    /// The decrypted decryption key
    bytes key_cipher = 3;

    /// The decryption iv
    bytes iv = 4;

    /// The amount
    int64 amount = 5;

    /// The invoice preimage hash.
    bytes preimage_hash = 6;

    /// The invoice
    string payment_request = 7;

    /// The seller node pubkey
    string pubkey = 8;

    /// The host of the seller lightning node
    string host = 9;

    /// The port of the seller lightning node
    int32 port = 10;
}
